simln-lib/src/lib.rs

```diff
@@ -1031,257 +1055,191 @@ impl<C: Clock + 'static> Simulation<C> {
+                _ = listener.clone()  => {
+                    return Ok::<(), SimulationError>(())
+                },
+                _ = async {} => {
+                    match heap.pop() {
+                        Some(Reverse(PaymentEvent {
+                            pubkey,
+                            absolute_time,
+                            destination,
+                            capacity
+                        })) => {
+                            let internal_now = SystemTime::now();
+                            let wait_time = match absolute_time.duration_since(internal_now) {
+                                Ok(elapsed) => {
+                                    // the output of duration_since is not perfectly round affecting the waiting time
+                                    // and as consequence the results are not deterministic; for this reason
+                                    // it is necessary to round the output.
+                                    Duration::from_secs(elapsed.as_secs_f64().round() as u64)
+                                },
+                                Err(_e) => {
+                                    Duration::from_secs(0)
                                 }
                             };
+                            let payload = payment_event_payloads.get(&pubkey).ok_or(SimulationError::PaymentGenerationError(
+                                    PaymentGenerationError(format!("executor {} not found", pubkey)),
+                            ))?;
+                            if let Some(c) = payload.executor.payment_generator.payment_count() {
+                                if c == payload.current_count {
+                                    log::info!( "Payment count has been met for {}: {c} payments. Stopping the activity.", payload.executor.source_info);
+                                    continue
+                                }
+                            }
+                            let node = nodes.get(&payload.executor.source_info.pubkey).ok_or(SimulationError::MissingNodeError(format!("Source node not found, {}", payload.executor.source_info.pubkey)))?.clone();
+
+                            // pe: produce events
+                            let pe_shutdown = shutdown.clone();
+                            let pe_clock = clock.clone();
+                            let source = payload.executor.source_info.clone();
+                            let pe_output_sender = output_sender.clone();
+                            // Only proceed with a payment if the amount is non-zero, otherwise skip this round. If we can't get
+                            // a payment amount something has gone wrong (because we should have validated that we can always
+                            // generate amounts), so we exit.
+                            let amount = payload.executor.payment_generator.payment_amount(capacity).map_err(SimulationError::PaymentGenerationError)?;
```

Comment:
Why not generate this along with the rest of the details we push onto the heap?

Doesn't necessarily have to change, just interested in motivation.